### ISSUE TITLE
Decide if any methods should be async

### ISSUE THREAD

**Issue by tomgilder** [tomgilder/routemaster#18]
Should anything else allow async and return a `Future`? For instance validators, guards and unknown page generators?

This could be added later but might be a mildly breaking change.

Mostly I don't think this is a good idea: routing requests generally want to be executed immediately, but I'm open to suggestions.

**Comment by theweiweiway** [comment]
i believe awaiting `pop` results  is a generally popular use case

**Comment by tomgilder** [comment]
Yeah, that's an interesting one. Currently Routemaster doesn't support pop results at all, and I'm not sure if it's easy to add them - because a push isn't a simple push of just one page in the stack, it can go anywhere.

**Comment by chimon2000** [comment]
Guards definitely make sense if you would like to perform validation inside of the guard.

**Comment by tomgilder** [comment]
I agree that it feels like guards should allow async - but I'm not sure how this should work in practice.

Whilst the async method is running, what should happen? 

Should the router lock and prevent other navigation?

What's the end use scenario here?

**Comment by tomgilder** [comment]
Here's the scenario that I find hard to figure out:

1. You have a `Guard` that takes an async method that takes 3 seconds to return
2. User navigates to that URL
3. What does the user see for those 3 seconds? What if they try to cancel the navigation?

But at the time, there are async methods that are quick and do I/O, which absolutely should be supported.

What about having an `AsyncGuard` that allowed you to show another page or spinner whilst it's loading?

**Comment by theweiweiway** [comment]
Something unrelated to this issue - but related to guards - is a `redirectTo` variable. Sometimes when you get redirected by a guard, you want to complete the guard and then redirect back to the initial page you were trying to get into. Usually on web - we can do this with a queryParameter. Not sure if on Flutter we can do a more integrated approach?

And in Flutter, it would probably be "popping" all the way back to the original page you were trying to access? Maybe there could be a flag like so:
```dart
Guard(
  redirect: true
)
```

Not sure how that would work tho..

**Comment by chimon2000** [comment]
> Here's the scenario that I find hard to figure out:
> 
> 1. You have a `Guard` that takes an async method that takes 3 seconds to return
> 2. User navigates to that URL
> 3. What does the user see for those 3 seconds? What if they try to cancel the navigation?
> 
> But at the time, there are async methods that are quick and do I/O, which absolutely should be supported.
> 
> What about having an `AsyncGuard` that allowed you to show another page or spinner whilst it's loading?

I like the idea of having an AsyncGuard, although I'm not sure it's the responsibility of the library to display loading in this scenario.  Using the Angular guard as the most flexible example:

```typescript
interface CanActivate {
  canActivate(route: ActivatedRouteSnapshot, state: RouterStateSnapshot): Observable<boolean | UrlTree> | Promise<boolean | UrlTree> | boolean | UrlTree
}
```

Just focusing on the first two scenarios since this is about async.

1. The developer returns an observable (stream) to show some route/UI while loading data and then replace that with some other route/UI.  
2. The developer returns a promise (future), blocking while data is loading then replacing that with some route/UI

I can see both as useful, maybe swapping out the stream for a ValueListenable, or just keeping it as is.

The Flutter examples that I have seen (vrouter, flutter_modular, qlevar_router) all handle the 2nd scenario.

**Comment by tomgilder** [comment]
> I'm not sure it's the responsibility of the library to display loading

**Comment by tomgilder** [comment]
> I'm not sure it's the responsibility of the library to display loading 

Absolutely, I'd never add this to the library - it'll be up to the user to show a loading spinner/page.

> The developer returns an observable (stream) to show some route/UI while loading data and then replace that with some other route/UI.

Interesting - how about an async method that can return multiple pages?

So you'd end up with something like this:

```dart
Stream<Page> buildRoute() async* {
  yield MaterialPage<void>(child: LoadingPage());

  if (await canShowPage()) {
    yield MaterialPage<void>(child: ContentPage());
  } else {
    yield MaterialPage<void>(child: BlockedPage());
  }
}
```

**Comment by chimon2000** [comment]
> > I'm not sure it's the responsibility of the library to display loading
> 
> Absolutely, I'd never add this to the library - it'll be up to the user to show a loading spinner/page.
> 
> > The developer returns an observable (stream) to show some route/UI while loading data and then replace that with some other route/UI.
> 
> Interesting - how about an async method that can return multiple pages?
> 
> So you'd end up with something like this:
> 
> ```dart
> Stream<Page> buildRoute() async* {
>   yield MaterialPage<void>(child: LoadingPage());
> 
>   if (await canShowPage()) {
>     yield MaterialPage<void>(child: ContentPage());
>   } else {
>     yield MaterialPage<void>(child: BlockedPage());
>   }
> }
> ```

Yes, I think this is potentially a very clean solution! Slightly tweaking to display redirects.

```dart
Stream<Page> buildRoute() async* {
  yield MaterialPage<void>(child: LoadingPage());

  if (await canShowPage()) {
    yield MaterialPage<void>(child: ContentPage());
  } else {
    yield Redirect('/login');
  }
}
```

**Comment by tomgilder** [comment]
Generally I'm heading towards thinking async in the router itself is a bad idea.

This is for the same reason Flutter doesn't allow `build` methods to be async - it needs to always know what to render. 

The above example can be done within the `Page` currently:

```dart
'/async': (_) => MaterialPage<void>(
      child: StreamWidgetBuilder(
        builder: (context) async* {
          yield Scaffold(body: Center(child: Text('Loading...')));

          if (await canShowPage()) {
            yield Scaffold(body: Center(child: Text('Success')));
          } else {
            Routemaster.of(context).replace('/');
          }
        },
      ),
    ),
```

Source for `StreamWidgetBuilder`:
<details>
  <summary>StreamWidgetBuilder</summary>

```dart
class StreamWidgetBuilder extends StatefulWidget {
  final Stream<Widget> Function(BuildContext) builder;
  final Widget initialChild;

  const StreamWidgetBuilder({
    Key? key,
    required this.builder,
    this.initialChild = const SizedBox(),
  }) : super(key: key);

  @override
  _StreamWidgetBuilderState createState() => _StreamWidgetBuilderState();
}

class _StreamWidgetBuilderState extends State<StreamWidgetBuilder> {
  late StreamSubscription<Widget> _subscription;
  late Widget _child;

  @override
  void initState() {
    super.initState();

    _child = widget.initialChild;
    _subscription = widget.builder(context).listen((newChild) {
      setState(() => _child = newChild);
    });
  }

  @override
  void dispose() {
    _subscription.cancel();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) => _child;
}
```
</details>

But I'd love any specific scenarios that would be unachievable without async.

**Comment by chimon2000** [comment]
I think there's not a huge difference between your initial proposal of the AsyncGuard and the above proposal besides the complexity in determining what to render since union types aren't a thing yet.  The above proposal is fine although the idea of calling a function in a `build` function _feels_ like a code smell.  Might be better to have a `Redirect` widget that accomplishes the same thing in `initState`.

**Comment by tomgilder** [comment]
> The above proposal works although the idea of calling a function in a `build` function _feels_ like a code smell. Might be better to have a `Redirect` widget that accomplishes the same thing in `initState`.

Great idea. But at the moment I've defined `Redirect` as a `Page` to use in the routing table 🤦 Argh!

**Comment by chimon2000** [comment]
Continuing your thought process on https://github.com/tomgilder/routemaster/issues/19#issuecomment-831226973, I really like the idea of deprecating guards in favor of just being able to return a Redirect or MaterialPage when declaring routes.

```dart
'/protected-route': (route) {
    if (!isLoggedIn()) return Redirect('/login');
    if (!canUserAccessPage) return Redirect('/no-access');
    return ProtectedPage();
},
```

In a more complex example, I could see wrapping your app in a bootstrap widget that would load authentication/authorization logic using a `FutureBuilder`/`StreamBuilder`before the routes are instantiated.

**Comment by tomgilder** [comment]
For now, I'm not going to implement any async behaviour. I want to get to a stable 1.0 release without the complexity of async.

I'll revisit this after that.

**Comment by tomgilder** [comment]
I've added a new example that shows how to handle several common async scenarios: logging in and deep linking.

See here: https://github.com/tomgilder/routemaster/tree/main/example/deep_linking

**Comment by bdremynoel** [comment]
Another scenario, pretty similar to login one, but subject to more changes during runtime.

Let's say we have a context, a house for example, urls could be like `/house/1/edit`.
Navigating to `/house/2/edit` could imply a reload of the house context:

-  data/logic: related permission vs. current user to edit, view or whatever.
-  UI: changing theme, etc.

This could lead to multiple unwanted checks in all sub-pages.

**Comment by tomgilder** [comment]
I've made an initial working prototype of this.

Here's an example of the current API:

```dart
'/async-page': (_) => AsyncPageBuilder(
      initialPage: MaterialPage(child: LoadingPage()),
      pageBuilder: () async* {
        final isLoggedIn = await appState.isLoggedIn();

        if (isLoggedIn) {
          yield Redirect('/login-page');
        } else {
          yield MaterialPage(child: HomePage());
        }
      },
    ),
```

It currently requires an `initialPage`. This could just be a totally blank page. 

It might be possible to avoid having an `initialPage`, and just wait until the first page is yielded from the `pageBuilder`.

If there is a Routemaster navigation event (such as the user entering a new URL), any further yielded pages are ignored to avoid confusing of background pages suddenly redirecting somewhere else.

Feedback welcome!

**Comment by tomgilder** [comment]
Okay, yeah, I found a way to get rid of `initialPage`... so it now works like this:

```dart
'/async-page': (_) => AsyncPageBuilder(
      pageBuilder: () async* {
        yield MaterialPage(child: LoadingPage());
        final isLoggedIn = await appState.isLoggedIn();

        if (isLoggedIn) {
          yield Redirect('/login-page');
        } else {
          yield MaterialPage(child: HomePage());
        }
      },
    ),
```

Only one API question left: does the `async*` need a way to know that the subscription has been cancelled? As in, the user has navigated to a new route? If so, how does it find out?

**Comment by bdremynoel** [comment]
Seems great!
About cancellation, I don't think we need to know about. This would result in bloated code anyway IMHO.

When cancelled, `pageBuilder` block is still fully processed (or am I wrong?), widget loaded but not attached to view hierarchy. So there should be no cascaded data loading.

What's the behaviour if nothing is generated? Expecting no route change here.

**Comment by tomgilder** [comment]
> About cancellation, I don't think we need to know about.

I think there are a handful of cases where this is useful. 

I'm currently thinking of just adding a parameter to the method:

```dart
'/async-page': (_) => AsyncPageBuilder(
      pageBuilder: (status) async* {
        yield MaterialPage(child: LoadingPage());
        final isLoggedIn = await appState.isLoggedIn();

        if (status.isCancelled) return;
        await doSomethingExpensive();
        if (status.isCancelled) return;
        await doAPICall();
        ...
      },
    ),
```

> What's the behaviour if nothing is generated? Expecting no route change here.

Good point. Currently it would do a navigation, but nothing would appear. A totally empty invisible page is added to the navigation stack, but the URL changes immediately. I agree it would be better if nothing happened at all; I'll look into it.

**Comment by bdremynoel** [comment]
This would open the door to bad conceptions. What if `doAPICall` throws an exception? Devs cannot recover the error from here, user is not informed, etc.

All the UI guidelines lead to provide the loading time period in the page itself with dedicated widget or variants of this.
We're navigating here, I vote for enforcing it to throw away as far as possible business cases.

**Comment by mdrideout** [comment]
I just wanted to post how I am currently using async routes with Routemaster in the event that there is an easier way I could be doing this now, or as another example for supporting an app startup in a cross-platform auth redirect situation.

## **Web Redirect Auth Considerations**
I am using OAuth2 cross platform via [App Auth ](https://pub.dev/packages/flutter_appauth)and the unofficial custom package for [app_auth_web](https://github.com/MaikuB/flutter_appauth/issues/70).

For native app, this uses an in-app browser style situation. But for **web**, the user is directed off app to the authentication portal (in this case **Azure AD**), and is then redirected to the Flutter web app html page **/oauthredirect.html** which captures the window.sessionStorage variables returned from the portal, before redirecting user back to the **initial route** of the app (such as '/').

**Async Processes Before Route Should Be Built**
During app startup, after the redirect back from the portal, we must authenticate against the portal to get the auth tokens. This must happen, before determining whether I should load my **loggedOutRouteMap** or **loggedInRouteMap**.

If the user hits the web refresh button, we'll do something similar, refreshing the user's authentication from session data saved to persistent storage. 

## Current Solution
- I am serving my RouteMaps from a Riverpod provider that watches my User state, and app initialization state.
- The Routemaster initial route is using FutureBuilder to show a loading screen while the startup process (future) is running, and if the User state is null, we'll show the loggedOutRouteMap. If the User is not null, we'll serve the loggedInRouteMap.

**FutureBuilder**
```dart
class InitialMaterialPageBuilder extends StatelessWidget {
  final Widget home;
  final Function(BuildContext) future;
  InitialMaterialPageBuilder({required this.home, required this.future});

  @override
  Widget build(BuildContext context) {
    return FutureBuilder(
      future: future(context),
      builder: (context, snapshot) {
        if (snapshot.hasError) {
          logger.e("App Init Failure: " + snapshot.error.toString());
          return AppInitFailScreen();
        }

        // When initAppFeatures future is done, load the main app Widget
        if (snapshot.connectionState == ConnectionState.done) {
          return home;
        }

        // Otherwise return loader
        return AppInitLoadingScreen();
      },
    );
  }
}
```

**The provider watching the user state and init state**
```dart
final routeMapProvider = Provider<RouteMap>((ref) {
  final User? _user = ref.watch(userObjectOnlyProvider);
  final bool _appInitComplete = ref.watch(globalStateProvider).appInitComplete;

  print("Rebuilding root map. InitState: $_appInitComplete");

  if (_user == null && _appInitComplete == true) {
    return RouteMap(
      onUnknownRoute: (_) => Redirect('/'),
      routes: {
        // '/' Initial route uses InitialPageBuilder() to complete app initialization functions at startup
        '/': (_) => MaterialPage(
              child: InitialMaterialPageBuilder(
                home: LoginScreen(),
                future: initAppFeatures,
              ),
            ),
      },
    );
  }

  return RouteMap(
    onUnknownRoute: (_) => MaterialPage(child: Error404Screen()),
    routes: {
      // '/' Initial route uses InitialPageBuilder() to complete app initialization functions at startup
      '/': (_) => MaterialPage(
            child: InitialMaterialPageBuilder(
              home: MonitoringDashboardScreen(),
              future: initAppFeatures,
            ),
          ),

      '/monitoring/job/:jobLogIndex': (route) =>
          MaterialPage(child: JobDetailScreen(jobIndexParam: route.pathParameters['jobLogIndex'])),

      // Business Dashboard
      '/business': (_) => MaterialPage(child: BusinessDashboardScreen()),
      '/business/settings': (_) => MaterialPage(child: BusinessSettingsWelcomeScreen()),
      '/business/job/:jobIndexParam': (route) =>
          MaterialPage(child: JobDetailScreen(jobIndexParam: route.pathParameters['jobIndexParam'])),

      // Machines
      '/machines': (_) => MaterialPage(child: MachinesListScreen()),
      '/machines/:machineId/dashboard': (route) => _machineExists(route.pathParameters['machineId'])
          ? MaterialPage(child: MachineDashboardScreen(machineId: route.pathParameters['machineId']))
          : MaterialPage(child: Error404Screen()),
    },
  );
});
```

**The root MaterialApp.router() widget, which watches the provider, to feed in a new RouteMap when the user state changes**
```dart
class RDGConnect extends ConsumerWidget {
  @override
  Widget build(BuildContext context, ScopedReader watch) {
    return FeatureDiscovery(
      child: MaterialApp.router(
        debugShowCheckedModeBanner: false,
        scaffoldMessengerKey: rootScaffoldMessengerKey,
        title: 'Roland DG Connect',
        theme: ConnectTheme.themeData(context),
        routerDelegate: RoutemasterDelegate(
          observers: [ConnectRouteObserver()],
          routesBuilder: (context) {
            RouteMap _routeMap = watch(routeMapProvider);
            return _routeMap;
          },
        ),
        routeInformationParser: RoutemasterParser(),
      ),
    );
  }
}
```

**Comment by jezsung** [comment]
What about making a guard just async function returning `Future` and exposing some navigation state via `Routemaster.of(context).state`?  The navigation state could contain some useful information about the current route navigating such as

from where
to where
isGuarding --> If the guard is an async function, the navigating would be this status
isTransitioning

I don't know if this is technically possible but just wanted to drop some of the ideas and suggesstions.

**Comment by mdrideout** [comment]
Here is an example of an easy way to create an async page that waits for a future to finish before loading the desired screen. It also handles displaying a loading screen while the future is running, and an error screen on error.

**Use Case**: in my case, for Flutter web, a user can enter the app from any internal screen and all screens require authentication. I would use a future that attempts to restore the user from persistent storage, and refresh their authentication tokens, before loading the desired screen. If the user does not exist or is expired, then a redirect can be handled with additional code.

### AsyncMaterialPage()
```dart
/// Async Material Page
class AsyncMaterialPage extends MaterialPage {
  final Widget screen;
  final Future future;
  AsyncMaterialPage({required this.screen, required this.future})
      : super(
          child: FutureBuilder(
            future: future,
            builder: (context, snapshot) {
              if (snapshot.hasError) {
                print("App Init Failure: " + snapshot.error.toString());
                return const AppInitFailScreen();
              }

              // When initAppFeatures future is done, load the main app Widget
              if (snapshot.connectionState == ConnectionState.done) {
                return screen;
              }

              // Otherwise return loader
              return const AppInitLoadingScreen();
            },
          ),
        );
}
```

An example of a `RouteMap()` where you can test a Future.delayed, and a Future.error scenario. Replaceable with any function that returns a future.
```dart
  return RouteMap(
    onUnknownRoute: (_) => const Redirect('/'),
    routes: {
      // '/' Initial route uses InitialPageBuilder() to complete app initialization functions at startup
      '/': (_) => AsyncMaterialPage(
            future: Future.delayed((Duration(seconds: 5))),
            // future: Future.error("Error future"),
            screen: LoginScreen(),
          ),
    },
  );
```

If your future will always be the same (like you have an app startup process that will need to potentially be run every time the app starts for any page visited - which can happen from Web), then you could omit the `future: ` field from `AsyncMaterialPage()`, and just wire the future directly into the FutureBuilder method in the constructor.
